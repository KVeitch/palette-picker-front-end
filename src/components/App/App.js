import React, { Component } from 'react';
import './App.scss';
import Color from '../Color/Color'
import InputForm from '../InputForm/InputForm'
import ProjectsContainer from '../ProjectsContainer/ProjectsContainer'




class App extends Component {
  constructor() {
    super()
    this.state= {
      user: {id:1, user_name:'Bob'},
      projects: [{project_name: 'Dining Room', project_id:1, 
        palettes:[
          {"id": 1,"palette_name": "Colors","project_id": 1,"color0": "6320ee","color1": "D5f2e3","color2": "999950","color3": "C7B446","color4": "C2B078"},
          {"id": 2,"palette_name": "Bright","project_id": 1,"color0": "786fa6","color1": "f19066","color2": "FFFF00","color3": "EDFF21","color4": "A98307"},
          {"id": 3,"palette_name": "Joker","project_id": 1,"color0": "211a1d","color1": "6320ee","color2": "D5f2e3","color3": "8075ff","color4": "8fc93a"},
          {"id": 4,"palette_name": "Russian Flat","project_id": 1,"color0": "f19066","color1": "786fa6","color2": "f19066","color3": "786fa6","color4": "574b90"},
          {"id": 5,"palette_name": "For Bill","project_id": 1,"color0": "16a085","color1": "2ecc71","color2": "AEA04B","color3": "E6D690","color4": "CDA434"}
        ]}],
      currentProjectId: null,
      color0: { hex:'4f4f4f', isLocked:false},
      color1: { hex:'1c6a77', isLocked:false},
      color2: { hex:'678b91', isLocked:false},
      color3: { hex:'95e6f4', isLocked:false},
      color4: { hex:'2dd0ed', isLocked:false}
    }
  }
  componentDidMount = () => {
    this.getStartingColors()
  }

  generateColor = () => {
    return Math.floor(Math.random()*16777215).toString(16);
  }

  getStartingColors = () => {
    this.setState({
      color0: { hex:this.generateColor(), isLocked:false},
      color1: { hex:this.generateColor(), isLocked:false},
      color2: { hex:this.generateColor(), isLocked:false},
      color3: { hex:this.generateColor(), isLocked:false},
      color4: { hex:this.generateColor(), isLocked:false}
    })
  }

  getColors = () => {
    let { color0, color1, color2, color3, color4 } = this.state;
    let colors =[color0, color1, color2, color3, color4];
    colors.forEach(color => {
      if (!color.isLocked) color.hex = this.generateColor(); 
    })
    this.setState({color0, color1, color2, color3, color4})
  }

  handleColorChange = (index, hex) =>{
    let { isLocked } = this.state[index];
    this.setState({ [index]: {hex: hex.substring(1), isLocked } })
  }

  toggleColorLock = (index) => {
    let { hex, isLocked} = this.state[index];
    console.log(hex,isLocked, index)
    isLocked = !isLocked;
    this.setState({[index]: {hex, isLocked } })
  }

  render = () =>{
    const { color0,color1,color2,color3,color4, projects } = this.state;
    const colors = [color0,color1,color2,color3,color4]
    const displayColors = colors.map((color, i)=>{
      colors.unshift(...colors.splice(-1));
      return <Color 
      colors={[...colors]}
      key={`color`+i}
      index={`color${4-i}`}
      handleColorChange={this.handleColorChange}
      toggleColorLock={this.toggleColorLock}/>
    })
    return (
      <div className="App">
        <nav><h1><span className="page__title">Some Title</span></h1>    Login</nav>
        <main>
          {displayColors}
        </main>
        <button type='button' 
        onClick={this.getColors}
        >Generate New Palette</button>
        <InputForm projects={projects} key={'inputForm'}/>
        {/* need function to edit and delete projects and palettes */}
        <ProjectsContainer projects={projects} key={'projectsContinaer'}/> 
      </div>
    )};

}

export default App;
