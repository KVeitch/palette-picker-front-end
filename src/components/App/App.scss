* {
  // border: solid 1px red;
  box-sizing:border-box;
  margin:0;
  padding:0;
}

.App {
  text-align: center;
}

nav {
  height: 5vh;
  display:flex;
  justify-content: space-between;
  padding: 10px;
}

main {
  display:flex;
  justify-content: space-between;
  padding:20px;
  border-radius: 15px;
  border:solid 1px black;
  margin:5vh 5vw;


}
