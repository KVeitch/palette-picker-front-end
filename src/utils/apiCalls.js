const baseUrl = 'https://api-palette-picker.herokuapp.com/api/v1'
// const baseUrl = 'http://localhost:3000/api/v1';
export const getAllPalettes = async () => {
  const response = await fetch(`${baseUrl}/palettes`);
  if (!response.ok) {
    console.error('Could not get palettes.');
  }
  const data = await response.json();
  return data;
};

export const getPaletteById = async id => {
  const response = await fetch(`${baseUrl}/palettes/${id}`);
  const data = response.json();
  return data;
};

export const getPalettesByColor = async color => {
  const response = await fetch(`${baseUrl}/palettes/search/${color}`);
  const data = response.json();
  return data;
};

export const getAllProjects = async () => {
  const response = await fetch(`${baseUrl}/projects`);
  if (!response.ok) {
    console.error('Could not get projects.');
  }
  const data = await response.json();
  return data;
};

export const getProjectById = async id => {
  const response = await fetch(`${baseUrl}/projects/${id}`);
  const data = response.json();
  return data;
};

export const postProject = async projectInfo => {
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(projectInfo)
  };

  const response = await fetch(`${baseUrl}/projects`, options);
  const data = response.json();
  return data;
};

export const postPalette = async paletteInfo => {
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(paletteInfo)
  };

  const response = await fetch(`${baseUrl}/palettes`, options);
  const data = response.json();
  return data;
};

export const deleteProject = async id => {
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  };
  const response = await fetch(`${baseUrl}/projects/${id}`, options);
  const data = response.json();
  return data;
};

export const deletePalette = async id => {
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  };
  const response = await fetch(`${baseUrl}/palettes/${id}`, options);
  const data = response.json();
  return data;
};

export const updateProject = async (id, update) => {
  const options = {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(update)
  };

  const response = await fetch(`${baseUrl}/projects/${id}`, options);
  const data = response.json();
  return data;
};

export const getUserProjects = async user_id => {
  const response = await fetch(`${baseUrl}/users/${user_id}/projects`);
  const data = await response.json();
  return data;
};

export const getProjectPalettes = async project_id => {
  const response = await fetch(`${baseUrl}/projects/${project_id}/palettes`);
  const data = await response.json();
  return data;
};
