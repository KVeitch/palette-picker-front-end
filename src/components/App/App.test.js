import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import { deleteProject } from '../../utils/apiCalls';

jest.mock('../../utils/apiCalls');

describe('App', () => {
  let wrapper;
  global.Math.random = jest
    .spyOn(global.Math, 'random')
    .mockImplementation(() => 0.123);

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('should match the snapshot with all the data passed in correctly', () => {
    wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call componentDidMount once', () => {
    let componentDidMountSpy = jest.spyOn(App.prototype, 
                           'componentDidMount');
    let component = shallow(<App/>);
  expect(componentDidMountSpy).toHaveBeenCalledTimes(1);
  });

  it('should call getStartingColors and setUserProjects from `componentDidMount()`', () => {
    const instance = wrapper.instance();
    jest.spyOn(instance, 'getStartingColors');
    jest.spyOn(instance, 'setUserProjects');
    instance.componentDidMount();
    expect(instance.getStartingColors).toHaveBeenCalledTimes(1); 
    expect(instance.setUserProjects).toHaveBeenCalledTimes(1); 
  });

  it('should generate random hexes', ()=>{
    let expected="1f7ced"
    let results = wrapper.instance().generateColor()
    expect(results).toEqual(expected)
  })

  it('should random colors',()=>{
    let expected = {
      color0: { hex: "1f7ced", isLocked: false },
      color1: { hex: "1f7ced", isLocked: false },
      color2: { hex: "1f7ced", isLocked: false },
      color3: { hex: "1f7ced", isLocked: false },
      color4: { hex: "1f7ced", isLocked: false }
    }
    wrapper.instance().getStartingColors();
    let results = {
      color0: wrapper.instance().state.color0,
      color1: wrapper.instance().state.color1,
      color2: wrapper.instance().state.color2,
      color3: wrapper.instance().state.color3,
      color4: wrapper.instance().state.color4
    }

    expect(results).toEqual(expected)
  })

  it('should make random color palettes',()=>{
    let expected = {
      color0: { hex: "1f7ced", isLocked: false },
      color1: { hex: "1f7ced", isLocked: false },
      color2: { hex: "1f7ced", isLocked: false },
      color3: { hex: "1f7ced", isLocked: false },
      color4: { hex: "1f7ced", isLocked: false }
    }
    wrapper.instance().getColors();
    let results = {
      color0: wrapper.instance().state.color0,
      color1: wrapper.instance().state.color1,
      color2: wrapper.instance().state.color2,
      color3: wrapper.instance().state.color3,
      color4: wrapper.instance().state.color4
    }
    expect(results).toEqual(expected)
  })

  it('should toggleColorLock', ()=>{
    let expected = { hex: "1f7ced", isLocked: true }
    wrapper.instance().toggleColorLock('color0', "#1f7ced")
    let results = wrapper.instance().state.color0
    expect(results).toEqual(expected)
  })

  it('should change the (state) current project', ()=>{
    let expected = 2
    wrapper.instance().state.projects = mockProjects
    wrapper.instance().setCurrentProject('Room2')
    let results = wrapper.instance().state.currentProjectId
    expect(results).toEqual(expected)
  })

  it('should have a user',()=>{
    let expected = { id: 1, user_name: 'Bob' }
    let results = wrapper.instance().state.user
    expect(results).toEqual(expected)
  })

  it('should call deleteProject and remove a project', () => {
    wrapper.instance().state.projects = mockProjects
    wrapper.instance().removeProject(1)
    let results = wrapper.instance().state.projects
    expect(deleteProject).toHaveBeenCalledTimes(1); 
    expect(results).toEqual(room2)
  });
  

});

//Data for tests
const mockProjects = [
  {
    project_name: 'Dining Room',
    id: 1,
    palettes: [
      {
        id: 1,
        palette_name: 'Colors',
        project_id: 1,
        color0: '6320ee',
        color1: 'D5f2e3',
        color2: '999950',
        color3: 'C7B446',
        color4: 'C2B078'
      },
      {
        id: 2,
        palette_name: 'Bright',
        project_id: 1,
        color0: '786fa6',
        color1: 'f19066',
        color2: 'FFFF00',
        color3: 'EDFF21',
        color4: 'A98307'
      },
      {
        id: 3,
        palette_name: 'Joker',
        project_id: 1,
        color0: '211a1d',
        color1: '6320ee',
        color2: 'D5f2e3',
        color3: '8075ff',
        color4: '8fc93a'
      },
      {
        id: 4,
        palette_name: 'Russian Flat',
        project_id: 1,
        color0: 'f19066',
        color1: '786fa6',
        color2: 'f19066',
        color3: '786fa6',
        color4: '574b90'
      },
      {
        id: 5,
        palette_name: 'For Bill',
        project_id: 1,
        color0: '16a085',
        color1: '2ecc71',
        color2: 'AEA04B',
        color3: 'E6D690',
        color4: 'CDA434'
      }
    ]
  },
  {
    project_name: 'Room2',
    id: 2,
    palettes: [
      {
        id: 1,
        palette_name: 'Colors',
        project_id: 2,
        color0: '6320ee',
        color1: 'D5f2e3',
        color2: '999950',
        color3: 'C7B446',
        color4: 'C2B078'
      },
      {
        id: 2,
        palette_name: 'Bright',
        project_id: 2,
        color0: '786fa6',
        color1: 'f19066',
        color2: 'FFFF00',
        color3: 'EDFF21',
        color4: 'A98307'
      },
      {
        id: 3,
        palette_name: 'Joker',
        project_id: 2,
        color0: '211a1d',
        color1: '6320ee',
        color2: 'D5f2e3',
        color3: '8075ff',
        color4: '8fc93a'
      },
      {
        id: 4,
        palette_name: 'Russian Flat',
        project_id: 2,
        color0: 'f19066',
        color1: '786fa6',
        color2: 'f19066',
        color3: '786fa6',
        color4: '574b90'
      },
      {
        id: 5,
        palette_name: 'For Bill',
        project_id: 2,
        color0: '16a085',
        color1: '2ecc71',
        color2: 'AEA04B',
        color3: 'E6D690',
        color4: 'CDA434'
      }
    ]
  }
];

const room2 = [
  {
    project_name: 'Room2',
    id: 2,
    palettes: [
      {
        id: 1,
        palette_name: 'Colors',
        project_id: 2,
        color0: '6320ee',
        color1: 'D5f2e3',
        color2: '999950',
        color3: 'C7B446',
        color4: 'C2B078'
      },
      {
        id: 2,
        palette_name: 'Bright',
        project_id: 2,
        color0: '786fa6',
        color1: 'f19066',
        color2: 'FFFF00',
        color3: 'EDFF21',
        color4: 'A98307'
      },
      {
        id: 3,
        palette_name: 'Joker',
        project_id: 2,
        color0: '211a1d',
        color1: '6320ee',
        color2: 'D5f2e3',
        color3: '8075ff',
        color4: '8fc93a'
      },
      {
        id: 4,
        palette_name: 'Russian Flat',
        project_id: 2,
        color0: 'f19066',
        color1: '786fa6',
        color2: 'f19066',
        color3: '786fa6',
        color4: '574b90'
      },
      {
        id: 5,
        palette_name: 'For Bill',
        project_id: 2,
        color0: '16a085',
        color1: '2ecc71',
        color2: 'AEA04B',
        color3: 'E6D690',
        color4: 'CDA434'
      }
    ]
  }
];