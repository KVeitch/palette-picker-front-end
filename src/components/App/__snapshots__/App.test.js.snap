// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with all the data passed in correctly 1`] = `
<div
  className="App"
>
  <nav>
    <h1>
      <span
        className="page__title"
      >
        Some Title
      </span>
    </h1>
        Login
  </nav>
  <main>
    <Color
      colors={
        Array [
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
        ]
      }
      handleColorChange={[Function]}
      index="color4"
      key="color0"
      toggleColorLock={[Function]}
    />
    <Color
      colors={
        Array [
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
        ]
      }
      handleColorChange={[Function]}
      index="color3"
      key="color1"
      toggleColorLock={[Function]}
    />
    <Color
      colors={
        Array [
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
        ]
      }
      handleColorChange={[Function]}
      index="color2"
      key="color2"
      toggleColorLock={[Function]}
    />
    <Color
      colors={
        Array [
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
        ]
      }
      handleColorChange={[Function]}
      index="color1"
      key="color3"
      toggleColorLock={[Function]}
    />
    <Color
      colors={
        Array [
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
          Object {
            "hex": "1f7ced",
            "isLocked": false,
          },
        ]
      }
      handleColorChange={[Function]}
      index="color0"
      key="color4"
      toggleColorLock={[Function]}
    />
  </main>
  <button
    onClick={[Function]}
    type="button"
  >
    Generate New Palette
  </button>
  <InputForm
    projects={
      Array [
        Object {
          "palettes": Array [
            Object {
              "color0": "6320ee",
              "color1": "D5f2e3",
              "color2": "999950",
              "color3": "C7B446",
              "color4": "C2B078",
              "id": 1,
              "palette_name": "Colors",
              "project_id": 1,
            },
            Object {
              "color0": "786fa6",
              "color1": "f19066",
              "color2": "FFFF00",
              "color3": "EDFF21",
              "color4": "A98307",
              "id": 2,
              "palette_name": "Bright",
              "project_id": 1,
            },
            Object {
              "color0": "211a1d",
              "color1": "6320ee",
              "color2": "D5f2e3",
              "color3": "8075ff",
              "color4": "8fc93a",
              "id": 3,
              "palette_name": "Joker",
              "project_id": 1,
            },
            Object {
              "color0": "f19066",
              "color1": "786fa6",
              "color2": "f19066",
              "color3": "786fa6",
              "color4": "574b90",
              "id": 4,
              "palette_name": "Russian Flat",
              "project_id": 1,
            },
            Object {
              "color0": "16a085",
              "color1": "2ecc71",
              "color2": "AEA04B",
              "color3": "E6D690",
              "color4": "CDA434",
              "id": 5,
              "palette_name": "For Bill",
              "project_id": 1,
            },
          ],
          "project_id": 1,
          "project_name": "Dining Room",
        },
      ]
    }
  />
  <ProjectsContainer
    projects={
      Array [
        Object {
          "palettes": Array [
            Object {
              "color0": "6320ee",
              "color1": "D5f2e3",
              "color2": "999950",
              "color3": "C7B446",
              "color4": "C2B078",
              "id": 1,
              "palette_name": "Colors",
              "project_id": 1,
            },
            Object {
              "color0": "786fa6",
              "color1": "f19066",
              "color2": "FFFF00",
              "color3": "EDFF21",
              "color4": "A98307",
              "id": 2,
              "palette_name": "Bright",
              "project_id": 1,
            },
            Object {
              "color0": "211a1d",
              "color1": "6320ee",
              "color2": "D5f2e3",
              "color3": "8075ff",
              "color4": "8fc93a",
              "id": 3,
              "palette_name": "Joker",
              "project_id": 1,
            },
            Object {
              "color0": "f19066",
              "color1": "786fa6",
              "color2": "f19066",
              "color3": "786fa6",
              "color4": "574b90",
              "id": 4,
              "palette_name": "Russian Flat",
              "project_id": 1,
            },
            Object {
              "color0": "16a085",
              "color1": "2ecc71",
              "color2": "AEA04B",
              "color3": "E6D690",
              "color4": "CDA434",
              "id": 5,
              "palette_name": "For Bill",
              "project_id": 1,
            },
          ],
          "project_id": 1,
          "project_name": "Dining Room",
        },
      ]
    }
  />
</div>
`;
